client/state/support/actions.js

```diff
@@ -4,10 +4,26 @@
  * Internal dependencies
  */
 
-import { SUPPORT_USER_ACTIVATE } from 'state/action-types';
+import {
+	SUPPORT_USER_ACTIVATE,
+	SUPPORT_SESSION_ACTIVATE,
+	SUPPORT_SESSION_EXPIRE,
+} from 'state/action-types';
 
 export function supportUserActivate() {
 	return {
 		type: SUPPORT_USER_ACTIVATE,
 	};
 }
+
+export function supportSessionActivate() {
+	return {
+		type: SUPPORT_SESSION_ACTIVATE,
+	};
+}
+
+export function supportSessionExpire() {
+	return {
+		type: SUPPORT_SESSION_EXPIRE,
+	};
+}
```

Comment:
it seems like we may not need two action types to handle what is essentially one state transition. we can provide these two action creators to provide the additional hinting to the callers, but under-the-hood I think there might be some benefit to keeping these transitions grouped into one thing. it depends really on whether we expect other things to be listening for these events and how deeply they want to respond; for instance, is it a pain to check `if ( wasActive && ! isActive )` vs. `if ( type === EXPIRING )`

```js
import { ACTIVE, EXPIRED } from 'lib/support/constants';

export const activate = () => ( {
	type: SUPPORT_SESSION_TOGGLE,
	nextState: ACTIVE,
} );

export const expire = () => ( {
	type: SUPPORT_SESSION_TOGGLE,
	nextState: EXPIRED,
} );
```

Reply:
There are three possibles states and transitions between them are one-way. The default state is No-session. No-session can transition to Active-session, and Active-session can transition to Expired-session. No other transitions are possible.

My intention is to explicitly support only those transitions, and it currently seems most intuitive to encode those as standalone action types. I don't expect other reducers to handle these actions, though it's always possible.

I'm not sure I follow regarding the pain of checking. Are you up for elaborating?

Reply:
my suggestion is a pragmatic remark about our general stat of action-type bloat. if we reframe this specific problem we're transitioning states, so a single action type can make due.

the internals of the action-types and reducers can handle all the semantic constraints while we can expose the API through action-creators that match what you originally wanted.

this isn't a big deal, just one simple way to reuse action types. for example, had we a bunch of places in our code where we were going to watch for the `SUPPORT_SESSION_ACTIVATE` action then it would give weight to having a specific action type for that. if we're only planning on supporting our largely-self-contained system and we're the only ones we want watching those actions then it shouldn't matter too much.

constraints on which transitions are possible should be roughly identical in terms of cost: either we trap on action type or we trap on the property of an action. that is, if we're `ACTIVE` we're either going to ignore another `_ACTIVATE` action _or_ we're going to ignore an action with `ACTIVE` as its `nextState`.

> I'm not sure I follow regarding the pain of checking. Are you up for elaborating?

just what I'm saying above. in some cases checking action types is easy because we can do it in a `switch` statement or in a simple `if/else` whereas checking the compound condition of action type + transition requires a few more bits of code

```js
const reducer = ( state = NO_SESSION, { type, nextState } ) => {
	if ( SUPPORT_SESSION_ALTER !== type ) {
		return state;
	}

	if (
		( state === ACTIVE && nextState === EXPIRED ) ||
		( state === EXPIRED && nextState === ACTIVE ) ||
		( state === NO_SESSION && nextState === ACTIVE )
	) {
		return nextState;
	}

	debug( 'invalid transition' );
	return state;
}
```

vs.

```js
const reducer = ( state = NO_SESSION, { type } ) => {
	switch ( type ) {
		case ACTIVE:
			return ( state === NO_SESSION || state === EXPIRED ) ? ACTIVE : ( debug( 'invalid' ), state );

		case EXPIRE:
			return state === ACTIVE ? EXPIRE : ( debug( 'invalid' ), state );

		default:
			return state;
	}
}
```

Ideally we'd have some pattern matching if indeed we wanted to maintain the constraints cleanly…

```elm
reducer state { type, nextState } =
	if SUPPORT_SESSION_CHANGE == type then
			case ( state, nextState ) of
				( ACTIVE, EXPIRED ) ->
					EXPIRED

				( EXPIRED, ACTIVE ) ->
					ACTIVE

				( NO_SESSION, ACTIVE ) ->
					ACTIVE

				_ ->
					state
	else
		state
```

Otherwise if we "trust the app" to not make invalid transitions it all collapses into something trivial either way.

```js
const unifiedReducer = ( state, { type = NO_SESSION, nextState } ) =>
	SUPPORT_SESSION_CHANGE === type
		? nextState
		: state;

const splitReducer = ( state = NO_SESSION, { type } ) => {
	switch ( type ) {
		case SUPPORT_SESSION_ACTIVATE:
			return ACTIVE;

		case SUPPORT_SESSION_EXPIRE:
			return EXPIRED;

		default:
			return state;
	}
}
```

Reply:
Thanks for elaborating. It was helpful, and I updated the state code to use a single action.